Core/src/IfcGeometryConverter/CurveConverter.h

```diff
@@ -2895,6 +2901,189 @@ namespace OpenInfraPlatform
+				}
+				template <>
+				carve::geom::vector<3> getDirectionOfCurve(const EXPRESSReference<typename IfcEntityTypesT::IfcPolynomialCurve>& polynomialCurve,
+					const typename IfcEntityTypesT::IfcNonNegativeLengthMeasure& parameter) const noexcept(false)
+				{
+					return getDirectionOfCurve(polynomialCurve, parameter * this->UnitConvert()->getLengthInMeterFactor());
+				}
+				template<>
+				carve::geom::vector<3> getDirectionOfCurve(const EXPRESSReference<typename IfcEntityTypesT::IfcPolynomialCurve>& polynomialCurve,
+					const double& parameter) const noexcept(false)
+				{				
+					//std::vector<typename IfcEntityTypesT::IfcReal> coefficientX, coefficientY, coefficientZ;
+					//std::vector<double> polynomialConstantX, polynomialConstantY, polynomialConstantZ;
+					int polynomialConstantCntX, polynomialConstantCntY, polynomialConstantCntZ;
+					double angleX, angleY, angleZ;
+					//get coefficients
+					if (polynomialCurve->CoefficientsX)
+					{
+						std::vector<double> polynomialConstantX;
+						std::vector<typename IfcEntityTypesT::IfcReal> coefficientX = polynomialCurve->CoefficientsX;
+						//convert to double
+						for (int i = 0; i < size(coefficientX); i++)
+						{
+							polynomialConstantX.push_back(double(coefficientX[i]));
+						}
+						polynomialConstantCntX = std::size(polynomialConstantX);
+						//calculate angle
+						angleX = SpiralUtils::AngleByAngleDeviationPolynomial(polynomialConstantX, polynomialConstantCntX, parameter);
+					}
+					if (polynomialCurve->CoefficientsY)
+					{
+						std::vector<double> polynomialConstantY;
+						std::vector<typename IfcEntityTypesT::IfcReal> coefficientY = polynomialCurve->CoefficientsY;
+						//convert to double
+						for (int i = 0; i < size(coefficientY); i++)
+						{
+							polynomialConstantY.push_back(double(coefficientY[i]));
+						}
+						polynomialConstantCntY = std::size(polynomialConstantY);
+						//calculate angle
+						angleY = SpiralUtils::AngleByAngleDeviationPolynomial(polynomialConstantY, polynomialConstantCntY, parameter);
+					}
+					if (polynomialCurve->CoefficientsZ)
+					{
+						std::vector<double> polynomialConstantZ;
+						std::vector<typename IfcEntityTypesT::IfcReal> coefficientZ = polynomialCurve->CoefficientsZ;
+						//convert to double
+						for (int i = 0; i < size(coefficientZ); i++)
+						{
+							polynomialConstantZ.push_back(double(coefficientZ[i]));
+						}
+						polynomialConstantCntZ = std::size(polynomialConstantZ);
+						//calculate angle
+						angleZ = SpiralUtils::AngleByAngleDeviationPolynomial(polynomialConstantZ, polynomialConstantCntZ, parameter);
+					}
+					
+					// calculate angle between two polynomial curves
+					//double angle = std::atan2(angleY, angleX);
+
+					if (polynomialConstantCntX>0 && polynomialConstantCntY>0)
+					{
+						double angle = std::atan2(angleY, angleX);
+						return carve::geom::VECTOR(std::cos(angle), std::sin(angle), 0.);
+
+					}

```

Comment:
Are you sure that you need to insert angleX and angleY here?
From a quick view it would feel more natural to insert pointX and pointY into atan to get an angle than using two angles to get another angle.
Please take a closer look here to see what you've than produces the results that you expect.

Reply:
I think you are right. 
I thought that we can find separatelly the angles for X and Y polynomial. Thats why I used the same function from SpiralUtils as for the clothoids. 
My suggestion to use the implementation from this sourse : https://firebirdsql.org/refdocs/langrefupd21-intfunc-atan2.html
Then we can use `atan2(sin(angleY), cos(angleX))`. 
See cf59b36

Reply:
I have tasted different cases and this is a result that presents the first data of vector `segmentDirections`

angle = atan2(std::sin(angleY),std::cos(angleX));
carve::geom::vector<3> v1 = carve::geom::VECTOR(std::cos(angle), std::sin(angle), 0. );

**[0]:{1.0000000000000000, 0.0000000000000000, 0.0000000000000000}- corresponds to point (0,0,0)**
[1]:{0.93699400295766744, 0.34934544282324143, 0.0000000000000000}
[2]:{-0.67306627244558215, 0.73958217453925257, 0.0000000000000000}
...
angle = atan2(y,x);
carve::geom::vector<3> v2 = carve::geom::VECTOR(std::cos(angle), std::sin(angle), 0. );

**[0]:{1.0000000000000000, 0.0000000000000000, 0.0000000000000000} - corresponds to point (0,0,0)**
[1]:{0.78609999999999047, 0.61795320999998504, 0.0000000000000000}
[2]: {1.2495999999999947, 1.5615001599999867, 0.0000000000000000}
...

It only says that we have similar values for different calculations. Interestingly, our rotation matrix uses only the first point (segmentDirections[0]) to detect the direction. I think this should be the next task to check over. 

P.S: I decided to place your suggestion about angle calculations.